pkg/server/api_conversion.go

```diff
@@ -0,0 +1,325 @@
+}
+
+var _ conversion.Factory = &CRConverterFactory{}
+
+// NewCRConverterFactory returns a CRConverterFactory that supports APIConversion-based conversions and the "none"
+// conversion strategy.
+func NewCRConverterFactory(apiConversionInformer apisinformers.APIConversionInformer) *CRConverterFactory {
+	return &CRConverterFactory{
+		nopConverter: conversion.NewNOPConverter(),
+		getAPIConversion: func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIConversion, error) {
+			return apiConversionInformer.Lister().Get(clusters.ToClusterAwareKey(clusterName, name))
+		},
+	}
+}
+
+// NewConverter returns the appropriate conversion.Converter based on the CRD. If the CRD identifies as for a wildcard
+// partial metadata request, the nop converter is used. Otherwise, it returns a CEL-based converter if there is an
+// associated APIConversion, a nop converter if the strategy is "none", or an error otherwise.
+func (f *CRConverterFactory) NewConverter(crd *apiextensionsv1.CustomResourceDefinition) (conversion.Converter, error) {
+	// Wildcard, partial metadata requests never need conversion
+	if strings.HasSuffix(string(crd.UID), ".wildcard.partial-metadata") {
+		return f.nopConverter, nil
+	}
+
+	// We have to return a deferredConverter here instead of an actual one because apiextensions-apiserver creates the
+	// converter once, when it creates serving info for a CRD. Because we aren't currently guaranteeing order (or cache
+	// freshness), there is always a chance that serving info is created for a CRD before an associated APIConversion
+	// exists and is present in the informer cache.
+	return &deferredConverter{
+		crd:          crd,
+		nopConverter: conversion.NewNOPConverter(),
+
+		celConverters: map[string]*celConverter{},
+
+		getAPIConversion: f.getAPIConversion,
+	}, nil
+}
+
+// deferredConverter implements conversion.Converter and determines on the fly what type of converter to use.
+type deferredConverter struct {
+	crd          *apiextensionsv1.CustomResourceDefinition
+	nopConverter conversion.Converter
+
+	// lock guards celConverters
+	lock sync.Mutex
+	// celConverters is a map from CRD UID to a celConverter.
+	celConverters map[string]*celConverter
+
+	getAPIConversion func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIConversion, error)
+}
+
+// Convert converts in to targetGV. If there is an APIConversion for this CRD, a CEL-based converter is used. Otherwise,
+// a nop converter is used.
+func (f *deferredConverter) Convert(in *unstructured.UnstructuredList, targetGV schema.GroupVersion) (*unstructured.UnstructuredList, error) {
+	converter, err := f.getConverter()
+	if err != nil {
+		return nil, fmt.Errorf("error getting converter: %w", err)
+	}
+
+	return converter.Convert(in, targetGV)
+}
+
+func (f *deferredConverter) getConverter() (conversion.Converter, error) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	converter, ok := f.celConverters[string(f.crd.UID)]
```

Comment:
If someone makes a mistake with a conversion definition and wants to update it, can this cache be cleared?

Reply:
You would have to create a new APIResourceSchema with a different name (don't have to change its spec though), a new APIConversion with a matching name, and update the APIExport to point at the new schema. We can look at making this easier in the future as a follow-up.